tutorials/haskell/snap-web-app/tutorial.md

```diff
@@ -0,0 +1,479 @@
+              </form>
+            </div>
+          </li>
+        </todos>
+      </ul>
+      <div>
+        <hr>
+        <div class="clearfix">
+          <!-- forms goes here  -->
+        </div>
+      </div>
+    </div>
+  </div>
+</apply>
+```
+Lets explain this a little bit. The `<apply template="base">` tag means this
+view uses a layout Heist template called `base.tpl`, you can have a closer look
+to it [here](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/snaplets/heist/templates/todos.tpl). The `<todos>` node implies everything it encloses will be repeated
+for every single Todo item. That comes from the Todo index request handler
+
+```haskell
+handleIndexTodo :: Handler App App ()
+handleIndexTodo = do
+  c <- gets _conn
+  todos <- liftIO $ Db.allTodos c
+  renderWithSplices "todos" $
+      "todos" ## spliceForTodos todos
+```
+
+which binds the `todos` tag with the splice obtained as result after calling `spliceForTodos` function.
+
+```haskell
+spliceForTodos :: [Db.Todo] -> SnapletISplice App
+spliceForTodos = mapSplices $ runChildrenWith . spliceForTodo
+```
+
+This `spliceForTodos` function uses another one called `spliceForTodo`, for
+creating a splice that binds Todo item values.
+
+```haskell
+spliceForTodo :: Db.Todo -> Splices (SnapletISplice App)
+spliceForTodo todo = do
+  "id"  ## textSplice . showAsText $ fromMaybe 0 ( Db.todoID todo )
+  "description"  ## textSplice $ Db.description todo
+  "completed"  ## textSplice . showAsText $ Db.completed todo
+  "notCompleted"  ## textSplice . showAsText $ not (Db.completed todo)
+  "successClass" ## textSplice $  if Db.completed todo
+                                    then "list-group-item-success"
+                                    else ""
+```
+
+Those values are used back in our [todos.tpl](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/snaplets/heist/templates/todos.tpl#L7-L22) Heist template as the
+`<description/>` tag or to build the **marks as done** link URL `/todos/${id}/complete/${notCompleted}`
+
+All this content is served through request handlers, which also control the
+application flow. This handlers do their job because they're binded to specific
+routes. Remember we talk about this while initializing the app?. Well, this is the route configuration provided to the application when calling the [`addRoutes`](http://hackage.haskell.org/package/snap-0.14.0.6/docs/Snap-Snaplet.html#v:addRoutes) function when it initializes .
+
+```haskell
+routes :: [(ByteString, Handler App App ())]
+routes = [ ("/",                                    handleIndexTodo)
+         , ("/todos/new",                           method GET handleIndexTodo
+                                                <|> method POST handleNewTodo)
+         , ("/todos/:todoId/complete/:completed",   method POST handleCompleteTodo)
+         , ("/todos/:todoId/delete",                method POST handleDeleteTodo)
```

Comment:
We should at least note that it's not a common practice anymore to configure your routes using strings. Something like `web-routes-boomerang` is preferred.